File: 83.删除排序链表中的重复元素.cpp
/*
 * @lc app=leetcode.cn id=83 lang=cpp
 *
 * [83] 删除排序链表中的重复元素
 */

// @lc code=start

// Definition for singly-linked list.

#include "iostream"
#include "set"
#include "map"

// struct ListNode {
//     int val;
//     ListNode *next;
//     ListNode() : val(0), next(nullptr) {}
//     ListNode(int x) : val(x), next(nullptr) {}
//     ListNode(int x, ListNode *next) : val(x), next(next) {}
// };

using namespace std;
class Solution {
public:
// 112
    ListNode* deleteDuplicates(ListNode* head) {
        set<int> m;
        if(!head || !head->next) {
            return head;
        }
        ListNode* current = head->next;
        ListNode* pre = head;
        m.insert(pre->val);
        while (current) {
            if(m.count(current->val)) {
                current = current->next;
                pre->next = current;
            }
            else {
                m.insert(current->val);
                pre = pre->next;
                current = current->next;
            }
        }
        return head;
    }
};
// @lc code=end

File: 83.删除排序链表中的重复元素.java
/*
 * @lc app=leetcode.cn id=83 lang=java
 *
 * [83] 删除排序链表中的重复元素
 */

// @lc code=start
// Definition for singly-linked list.
// public class ListNode {
//     int val;
//     ListNode next;
//     ListNode() {}
//     ListNode(int val) { this.val = val; }
//     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
// }

class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if(head == null || head.next == null){
            return head;
        }
        head.next = deleteDuplicates(head.next);
        if(head.val == head.next.val) head = head.next;
        return head;
        // if(head==null){
        //     return null;
        // }
        // Set<Integer> set=new HashSet<Integer>();
        // set.add(head.val);
        // ListNode q=head;
        // ListNode p=head.next;
        // while(p!=null){
        //     if(set.contains(p.val)){
        //         p=p.next;
        //         q.next=p;
        //     }else{
        //         set.add(p.val);
        //         q=q.next;
        //         p=p.next;
        //     }
        // }
        // return head;
    }
}
// @lc code=end

